### PR TITLE
Develop oss parenting

### PR DESCRIPTION
**Pull request by todderasesareddot**
Unifying ways to parent a child.  

* Now, child.setParent() and parent.addChild() will always have the same result.

* child.setParent(None) will remove child from parent's children list.

* addChild() renames an object automatically if there is a name clash at the same level.  Now, it will set the name back to the intended name if called again and there is no clash.  (added _originalName)

* getNameDecoration() returns the component side.  Without this, we cannot parent two leg joints under a pelvis joint when the naming is done according to the config.  If this method is blank, what is the intention?

### CODE REVIEW

File: Python/kraken/core/objects/object_3d.py [hunk 1]
```diff
@@ -272,7 +272,7 @@ def setName(self, name):
             bool: True if successful.
 
         """
```
Comment (EricTRocks) [inline Python/kraken/core/objects/object_3d.py]: Need an empty space after doc string. Apply to rest of code changes as well.

**Comment by EricTRocks** [comment]
Storing an _originalName attr and having the name magically change when parented under a different object that wouldn't cause a name clash with the original name isn't very intuitive. If I code something to add a locator under another one and a name clash exists, it'll get an "01" appended to it. If I query the children of that parent I'll see that object has been renamed because of a clash. So now I see the object listed as "myObject01". Then I re-parent it and try to find the object by name it won't exist with that same name and magically changed without a real cause.

The cause of the 01 being added was that there was a name clash during that `addChild()` call. Simply re-parenting when there is no clash shouldn't cause a name change. This is counter intuitive to how most if not all other apps work. Maya and Softimage for example won't do this.

It really is up to the user to ensure that names don't clash when running operations such as parenting and such. This is the only issue I have with this pull request.

**Comment by todderasesareddot** [comment]
I had a feeling you were going kick that original name aspect back.  :)  I have gone back and forth, myself.

I'll agree that restoring an original name is not very intuitive, but magically changing it upon parenting is not very convenient either.

To play Devil's advocate with your example.  Conversely, as it stands if you add a locator "myObject"  to another one, and the name does not clash, you will obviously not get an "01" appended to it and all is good.  So we see the object listed as "myObject".  Then you re-parent it to a new parent later that already has a "myObject" and it magically gets renamed to "myObject01".  Is that not just as problematic?  In fact, if you try to find the object by name after that, it will actually return the wrong object altogether.  

If the system magically renames nodes based on the addChild() method, you need to re-assess any list of object names anyway if you are doing anything with object string names and not object references.  If that's the case, why not name it back to what the author wanted it to be?

I would argue that it is reasonable for DCCs to automagically rename objects upon creation or reparenting because those actions can be and are often atomic operations and names/namespaces is the primary mode of identifying and working with objects.  Whereas Kraken is very different and rarely should any object be handled by name.  Especially since there is a rule-based naming mechanism during the Kraken build, it seems to only complicate things by renaming objects internally.

In my opinion, it would be much better for Kraken to keep the original intended name for as long as possible (i.e. until getBuildName() produces a clash).  That is, don't rename internally at all, just guarantee unique DCC node names during build.  Of course, authors should be cognizant about naming, but I really don't see the benefit of appending an "01" to the kraken object name.  In some cases like ours it actually screws things up.  If you have a component that has objects with the same name at the same level, chances are it was intentional (like my example of a two "leg" joint objects parented under a "pelvis".  They have different locations and I know that the DCC build name will resolve properly later on.

**Comment by todderasesareddot** [comment]
Also, in this pull request, getNameDecoration() now returns the  component side.  This effectively solves our OSS name clash issue (left and right "leg" joint objects parented under "pelvis", but really what it's doing is bypassing the renaming done by addChild() and allowing two nodes to exist with the same name and same parent.

Since getNameDecoration() returned nothing prior, what was the intent of this function?

**Comment by EricTRocks** [comment]
To clarify on this stance. There are a number of reasons to have objects auto-rename when being added under a parent with a node named the same. First one is where Components come in to play. They sub-class off of the low level classes and they also inherit the logic for auto-renaming.

In the UI, the user can drop in 2 arm components. They 2nd arm component needs to be renamed as to not clash. Otherwise the user won't understand which component is which and won't be able to easily distinguish the 2 arms.

If we don't rename single objects as they are added to the graph, then it will be very difficult to debug which objects were renamed due to name clashes. An example is if you add 2 inputs to a component with the same name. This would be legal up until build time and would also cause errors. This has already come up in some edge cases with Operators and if we allow this all the way up until build time, connection look ups won't be able to be done.

```Python
myInput1 = ScalarAttribute('myInput1', parent=attrParent)
myInput2 = ScalarAttribute('myInput1', parent=attrParent)

# At build time when iterating over the '_attributes' list and finding the input by name you may end up with the wrong item
```

There would have to be a lot more error handling there is currently all of the build system if we don't ensure that each object has a unique name.

## More work needed
That said, there are some cases where having objects within the same component that share a name is helpful. A great example is a face rig where you have controls on both sides of the face but only want to have 1 component that contains both. We need to provide a more elegant solution for sure.

**Comment by todderasesareddot** [comment]
Right on.  I appreciate the desire to have unique names.

The reason this came up this is because we ran into road blocks with the current implementation.  We had a number of cases where we needed two objects to exist under the parent with the same name.  You mention the face rig, but we have others, too.  Generally, the only reason they have the same name is because the assignment of their full names are deferred until the build step when the config comes into play.  For example, the side token is not added until the build at which point the names will no longer be the same.  Appending a number throws a monkey wrench in the works.

As far as attributes, I would consider those to be a different case, though.  They are not going to be renamed in the builder so clearly they need to be unique if they have the same parent.

Same with, components.  They are not going to be renamed in the builder like 3D objects and clearly need distinct names.

The _originalName attribute/behavior was just something I threw in here to try and we don't rely on this now that we have getNameDecoration() returns info that disambiguates our nodes. 

So, the first two bullet points might be good to integrate, though.  They solidify any parenting calls by making sure the same action is done no matter what way you go about parenting.

**Comment by EricTRocks** [comment]
Yes the parenting improvements are welcome and if you resubmit in a separate pull request we'll get those merged in right away.

**Pull request closed without merging**